File: src/execgo_playground/benchmarks/report.py
from __future__ import annotations

from collections import defaultdict

from ..models import BenchmarkResult


def render_run_summary(result: BenchmarkResult) -> str:
    metrics = result.metrics
    verdict = "PASS" if result.verdict.passed else "FAIL"
    return "\n".join(
        [
            f"# Run Summary: {result.run_id}",
            "",
            f"- Framework: `{result.framework}`",
            f"- Scenario: `{result.scenario_id}`",
            f"- Mode: `{result.mode}`",
            f"- Chaos: `{result.chaos_profile}`",
            f"- Verdict: `{verdict}`",
            f"- Wall time: `{metrics.wall_time_ms} ms`",
            f"- Tasks: `{metrics.task_count}` across `{metrics.stage_count}` stage(s)",
            f"- Runtime failures: `{metrics.runtime_failure_count}`",
            f"- Artifact hash match: `{metrics.artifact_hash_match}`",
        ]
    )


def render_benchmark_summary(results: list[BenchmarkResult]) -> str:
    grouped: dict[str, list[BenchmarkResult]] = defaultdict(list)
    for result in results:
        grouped[result.framework].append(result)

    lines = [
        "# Benchmark Summary",
        "",
        "| Framework | Runs | Pass Rate | Avg Wall Time (ms) |",
        "| --- | ---: | ---: | ---: |",
    ]
    for framework, framework_results in sorted(grouped.items()):
        total = len(framework_results)
        passes = sum(1 for item in framework_results if item.verdict.passed)
        avg = sum(item.metrics.wall_time_ms for item in framework_results) / max(total, 1)
        lines.append(f"| {framework} | {total} | {passes / max(total, 1):.2f} | {avg:.0f} |")
    return "\n".join(lines) + "\n"

File: src/execgo_playground/benchmarks/report.ts
import type { BenchmarkResult } from "../models.js";

export function renderRunSummary(result: BenchmarkResult): string {
  const metrics = result.metrics;
  const verdict = result.verdict.passed ? "PASS" : "FAIL";
  return [
    `# Run Summary: ${result.run_id}`,
    "",
    `- Framework: \`${result.framework}\``,
    `- Scenario: \`${result.scenario_id}\``,
    `- Mode: \`${result.mode}\``,
    `- Chaos: \`${result.chaos_profile}\``,
    `- Verdict: \`${verdict}\``,
    `- Wall time: \`${metrics.wall_time_ms} ms\``,
    `- Tasks: \`${metrics.task_count}\` across \`${metrics.stage_count}\` stage(s)`,
    `- Runtime failures: \`${metrics.runtime_failure_count}\``,
    `- Artifact hash match: \`${metrics.artifact_hash_match}\``,
  ].join("\n");
}

export function renderBenchmarkSummary(results: BenchmarkResult[]): string {
  const grouped = new Map<string, BenchmarkResult[]>();
  for (const result of results) {
    grouped.set(result.framework, [...(grouped.get(result.framework) ?? []), result]);
  }
  const lines = [
    "# Benchmark Summary",
    "",
    "| Framework | Runs | Pass Rate | Avg Wall Time (ms) |",
    "| --- | ---: | ---: | ---: |",
  ];
  for (const [framework, frameworkResults] of [...grouped.entries()].sort(([left], [right]) => left.localeCompare(right))) {
    const total = frameworkResults.length;
    const passes = frameworkResults.filter((item) => item.verdict.passed).length;
    const avg = frameworkResults.reduce((sum, item) => sum + item.metrics.wall_time_ms, 0) / Math.max(total, 1);
    lines.push(`| ${framework} | ${total} | ${(passes / Math.max(total, 1)).toFixed(2)} | ${avg.toFixed(0)} |`);
  }
  return `${lines.join("\n")}\n`;
}
